### PR TITLE
refactor(bo): simplify API endpoints, move bl to central lib folder

### PR DESCRIPTION
**Pull request by lucacavallaro**
#### List of Changes

1. moved all the code from `app/**/_lib/` to `@/lib`
2. created a new module `@/lib/api-keys` with exposed function for the front-end (`index.ts`), and backing service integration (`cosmos.ts`, `apim.ts`)
3. simplified (?) `ApiKey` entity
4. simplified `PATCH` endpoint to update an `ApiKey`, now it replaces the entire collection

#### Motivation and Context

this structure better suits with the front-end (that reuses part of this code)

### CODE REVIEW

File: apps/io-sign-backoffice-app/src/lib/api-keys/use-cases.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+export async function getApiKeyWithSecret(
+  id: string,
+  institutionId: string
+): Promise<ApiKeyWithSecret> {
+  try {
+    const apiKey = await getApiKey(id, institutionId);
+    return exposeApiKeySecret(apiKey);
+  } catch (e) {
+    console.log(e);
```
Comment (silvicir) [inline apps/io-sign-backoffice-app/src/lib/api-keys/use-cases.ts:79]: remove console.log

File: apps/io-sign-backoffice-app/src/lib/api-keys/apim.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  apiKey: ApiKey
+): Promise<string> {
+  return getApimClient().createSubscription(apiKey.id, apiKey.displayName);
+}
+
+export function deleteApiKeySubscription(apiKey: ApiKey): Promise<void> {
+  return getApimClient().deleteSubscription(apiKey.id);
+}
+
+export async function exposeApiKeySecret(
```
Comment (silvicir) [inline apps/io-sign-backoffice-app/src/lib/api-keys/apim.ts:17]: maybe we can think about removing `exposeApiKeySecret` and adding a `getSecret` like this
```
export async function getSecret(apiKey: ApiKey) {
  return getApimClient().getSecret(apiKey.id);
}
```
(similar to the other ones) and moving the other logic to `use-cases.ts` (either as a constructor or a generic function)

File: apps/io-sign-backoffice-app/src/app/api/institutions/[institution]/api-keys/[api-key]/route.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  }
+  try {
+    const body = await request.json();
+    const {
+      0: { path, value: newValue },
+    } = bodySchema.parse(body);
+    await upsertApiKeyField(
+      params["api-key"],
+      params.institution,
+      path === "/cidrs" ? "cidrs" : "testers",
```
Comment (gunzip) [inline apps/io-sign-backoffice-app/src/app/api/institutions/[institution]/api-keys/[api-key]/route.ts:61]: can you add a comment here? it's not very clear what's happening

File: apps/io-sign-backoffice-app/src/lib/api-keys/cosmos.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  displayName: string
+): Promise<boolean> {
+  const apiKeys = getApiKeys(institutionId, displayName);
+  const item = await apiKeys.next();
+  return !item.done;
+}
+
+export async function insertApiKey(apiKey: ApiKey): Promise<void> {
+  const { cosmosContainerName } = getCosmosConfig();
+  await getCosmosContainerClient(cosmosContainerName).items.create(apiKey);
```
Comment (gunzip) [inline apps/io-sign-backoffice-app/src/lib/api-keys/cosmos.ts:49]: no need to wrap any error here? `("Unable to insert key...")`

File: apps/io-sign-backoffice-app/src/lib/api-keys/use-cases.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        "such name already exists for the institution"
+      );
+    }
+    const apiKey = ApiKey(payload);
+    await createApiKeySubscription(apiKey);
+    try {
+      await insertApiKey(apiKey);
+    } catch (e) {
+      await deleteApiKeySubscription(apiKey);
+      throw new Error("unable to create the API key", { cause: e });
```
Comment (gunzip) [inline apps/io-sign-backoffice-app/src/lib/api-keys/use-cases.ts:50]: maybe you can move this error into `insertApiKey`?

**Review by gunzip — APPROVED** [review]

File: apps/io-sign-backoffice-app/src/lib/api-keys/use-cases.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      throw new ApiKeyAlreadyExistsError(
+        "such name already exists for the institution"
+      );
+    }
+    const apiKey = ApiKey(payload);
+    await createApiKeySubscription(apiKey);
+    try {
+      await insertApiKey(apiKey);
+    } catch (e) {
+      await deleteApiKeySubscription(apiKey);
```
Comment (silvicir) [inline apps/io-sign-backoffice-app/src/lib/api-keys/use-cases.ts:47]: Before there was a `throw new Error("unable to create the API key", { cause: e });`. As it is now, it returns an id even if there is an error

**Pull request merged**